File: CONCEPTS/reshaping.py
import pandas as pd
import sqlite3
import os

def dump(df):
    print(df)
    print()
    return df

path_to_database = os.path.join(os.environ["EXAMPLES"], "Boxscores.db")
con = sqlite3.connect(path_to_database)

# Bring in player and team names
query = """
    SELECT b.*, p.PLAYER_NAME, t.TEAM_NAME
    FROM Boxscores b
    LEFT JOIN Players p ON b.PLAYER_ID = p.PLAYER_ID
    LEFT JOIN Teams   t ON b.TEAM_ID   = t.TEAM_ID
    ORDER BY b.GAME_ID, b.TEAM_ID
"""
shooting = pd.read_sql(query, con)
shooting = shooting.drop(columns=["PLAYER_ID", "TEAM_ID"])
con.close()

print(shooting)

shooting_summary = (shooting

    # pivot_longer: stack the 6 stat columns into stat_type + count
    .melt(
        id_vars    = ["GAME_ID", "PLAYER_NAME", "TEAM_NAME"],
        value_vars = ["FGM", "FGA", "FG3M", "FG3A", "FTM", "FTA"],
        var_name   = "stat_type",
        value_name = "count"
    )
    .pipe(dump)

    # mutate: split stat_type into shot_type and made_attempt
    .assign(
        shot_type    = lambda df: df["stat_type"].str.replace(r"(M|A)$", "", regex=True),
        made_attempt = lambda df: df["stat_type"].str.extract(r"(M|A)$")
    )
    .drop(columns="stat_type")
    .pipe(dump)

    # pivot_wider: spread M and A into separate columns
    .pivot_table(
        index   = ["GAME_ID", "PLAYER_NAME", "TEAM_NAME", "shot_type"],
        columns = "made_attempt",
        values  = "count",
        aggfunc = "sum"
    )
    .reset_index()
    .pipe(dump)

    # group_by + summarise
    .groupby(["PLAYER_NAME", "TEAM_NAME", "shot_type"])
    .apply(lambda g: pd.Series({
        "attempts": g["A"].sum(),
        "pct":      g["M"].sum() / g["A"].sum() * 100
    }), include_groups=False)
    .reset_index()
)

print(shooting_summary)

File: CONCEPTS/reshaping.R
library(tidyverse)
library(RSQLite)


dump <- function(x) {
  print(x)
  cat("\n")
  return(invisible(x))
}

# Connect to database with error handling
path_to_database <- file.path(Sys.getenv("EXAMPLES"), "Boxscores.db")
 
con <- dbConnect(RSQLite::SQLite(), path_to_database)

# Bring into Boxscores the player and team names
query <- tbl(con, "Boxscores")                     |>
  left_join(tbl(con, "Players"), by = "PLAYER_ID") |>
    left_join(tbl(con, "Teams"), by = "TEAM_ID")   |>
        arrange(GAME_ID, TEAM_ID)                  |>
          select(-PLAYER_ID, -TEAM_ID)             |>
            as_tibble() 
           

# Run the query
shooting <- query |>
  collect()

print(shooting)

dbDisconnect(con)

shooting_summary <- shooting |>

  pivot_longer(
    cols      = c(FGM, FGA, FG3M, FG3A, FTM, FTA),
    names_to  = "stat_type",
    values_to = "count"
  ) |> dump() |>

  mutate(
    shot_type    = str_remove(stat_type, "(M|A)$"),
    made_attempt = str_extract(stat_type, "(M|A)$")
  ) |> dump() |>

  select(-stat_type) |>

  pivot_wider(
    names_from  = made_attempt,
    values_from = count,
    values_fn   = sum
  ) |> dump() |>
  
  group_by(PLAYER_NAME, TEAM_NAME, shot_type) |>
  summarise(
    attempts = sum(A),
    pct      = sum(M) / sum(A) * 100,
    .groups  = "drop"
  )

print(shooting_summary)




